src/main/java/org/springframework/retry/annotation/RecoverAnnotationRecoveryHandler.java

```diff
@@ -173,27 +174,34 @@ public void doWith(Method method) throws IllegalArgumentException, IllegalAccess
 				if (recover == null) {
 					recover = findAnnotationOnTarget(target, method);
 				}
-				if (recover != null && method.getReturnType().isAssignableFrom(failingMethod.getReturnType())) {
-					Class<?>[] parameterTypes = method.getParameterTypes();
-					if (parameterTypes.length > 0 && Throwable.class.isAssignableFrom(parameterTypes[0])) {
-						@SuppressWarnings("unchecked")
-						Class<? extends Throwable> type = (Class<? extends Throwable>) parameterTypes[0];
-						types.put(type, method);
-						RecoverAnnotationRecoveryHandler.this.methods.put(method,
-								new SimpleMetadata(parameterTypes.length, type));
-					}
-					else {
-						RecoverAnnotationRecoveryHandler.this.classifier.setDefaultValue(method);
-						RecoverAnnotationRecoveryHandler.this.methods.put(method,
-								new SimpleMetadata(parameterTypes.length, null));
+				if (recover != null && failingMethod.getGenericReturnType() instanceof ParameterizedType) {
+					if (method.getGenericReturnType().equals(failingMethod.getGenericReturnType())) {
```

Comment:
While this fixes the reported issue, unfortunately, I think it needs to be more sophisticated; we need to examine the generic parameter(s) `Map<String, Foo>` should match `Map<String, Bar>` if `Bar` is a subclass of `Foo`.

Reply:
Created a method `isParameterizedTypeAssignable` to verify the `actualTypeArguments` with `isAssignableFrom` check. Please review and let me know if this is what we expect.

Reply:
@garyrussell I have made the changes as requested. Can you please review?